utils/p4runtime_lib/convert.py

```diff
@@ -43,6 +43,21 @@ def encodeIPv4(ip_addr_string):
 def decodeIPv4(encoded_ip_addr):
     return socket.inet_ntoa(encoded_ip_addr)
 
+def matchesIPv6(ip_addr_string):
+    if not ipv6_pattern.match(ip_addr_string):
```

Comment:
Since `ipv6_pattern` is not defined, it should no longer be used, either.  You should try testing this code before submitting it for review, please.

Reply:
It is very much not a trivial thing to come up with a regex that matches all legal IPv6 address strings, and only legal IPv6 address strings, primarily because of the ability to use `::` in the middle of IPv6 addresses to replace consecutive strings of 0s.  See https://en.wikipedia.org/wiki/IPv6_address for examples (search for "::" to find the examples).

I am pretty sure that there are one or more built-in Python functions for converting such strings into IPv6 addresses, and they have been heavily tested by others to accept these things.  We should leverage those rather than trying to create a new implementation of all of those cases, unless your goal is to create yet another test for those built-in Python implementations.

Reply:
But why we have defined separate regex pattern for `mac_pattern` & `ip_pattern` here?

https://github.com/p4lang/tutorials/blob/f68e6f01f7303f17917327c96901da1f6ca6c7b7/utils/p4runtime_lib/convert.py#L26

https://github.com/p4lang/tutorials/blob/f68e6f01f7303f17917327c96901da1f6ca6c7b7/utils/p4runtime_lib/convert.py#L36

Reply:
As far as I can tell from a quick review, the mac_pattern is 100% accurate, with no false positives or negatives.  The ip_pattern regex does match some strings that are not legal IPv4 addresses, e.g. it will match `999.999.999.999`.

Getting even half-way close with IPv6 addresses with a regex is much harder.

Reply:
Ex.
```
import socket

def check_ipv6(n):
    try:
        socket.inet_pton(socket.AF_INET6, n)
        return True
    except socket.error:
        return False
```

Would it be possible to use try/except with the above thing, as you indicated in earlier comments?

Reply:
- So far, I have given corner test cases and a few more cases for utilising `socket.inet_pton` and getting accurate results.
- Tested cases compiler link is [here](https://onecompiler.com/python/429xku7pn)

Reply:
When I attempt to click on that link you gave for the test cases, I see a 401 error in my browser.  Can you put them some place others can view on the public Internet, please?

Reply:
You can look at [here](https://www.online-python.com/TwxqUKPmXJ) however, I'm not sure why it gave you a 401 error.